src/state/wallet/hooks.ts

```diff
@@ -67,23 +75,30 @@ export function useTokenBalancesWithLoadingIndicator(
 
   const anyLoading: boolean = useMemo(() => balances.some(callState => callState.loading), [balances])
 
-  return [
-    useMemo(
-      () =>
-        address && validatedTokens.length > 0
-          ? validatedTokens.reduce<{ [tokenAddress: string]: TokenAmount | undefined }>((memo, token, i) => {
-              const value = balances?.[i]?.result?.[0]
-              const amount = value ? JSBI.BigInt(value.toString()) : undefined
-              if (amount) {
-                memo[token.address] = TokenAmount.fromRawAmount(token, amount)
-              }
-              return memo
-            }, {})
-          : {},
-      [address, validatedTokens, balances],
-    ),
-    anyLoading,
-  ]
+  const balanceResult: { [key: string]: TokenAmount } = useMemo(
+    () =>
+      address && validatedTokens.length > 0
+        ? validatedTokens.reduce<{ [tokenAddress: string]: TokenAmount | undefined }>((memo, token, i) => {
+            const value = balances?.[i]?.result?.[0]
+            const amount = value ? JSBI.BigInt(value.toString()) : undefined
+            if (amount) {
+              memo[token.address] = TokenAmount.fromRawAmount(token, amount)
+            }
+            return memo
+          }, {})
+        : EMPTY_BALANCE,
+    [address, validatedTokens, balances],
+  )
+
+  // `balanceResult` was calculated base on `balances`, when `balances` changes, `balanceResult` recalculated
+  // again and return new instance of the result.
+  // But sometimes (most time likely), new result and old result are same, but have different instance.
+  // Below we are going to cache it, so if new result deep equals to old result, old result's instance will be use instead
+  // This cache helps hooks which calling this hooks and depend on this result don't have to calculating again with new dependency changed
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const balanceResultCached = useMemo(() => balanceResult, [stringifyBalance(balanceResult)])
+
+  return useMemo(() => [balanceResultCached, anyLoading], [anyLoading, balanceResultCached])
```

Comment:
Is useMemo needed here? as we eventually spread the returned array 🤔

Reply:
yea, not needed